File: build_data.ts
import * as fs from 'fs/promises';
import * as path from 'path';
import { parse } from 'csv-parse/sync';

const SHEET_ID = "1fnay67_hc7hU1BRAhbcfnQW5TQGReXGWpn0vIa_dkMY";
const CSV_URL = `https://docs.google.com/spreadsheets/d/${SHEET_ID}/export?format=csv`;

const PLANTS_JSON_PATH = "data/plants.json";
const DATA_DIR = "data";

async function main() {
    console.log(`Downloading CSV from ${CSV_URL}...`);
    const response = await fetch(CSV_URL);
    if (!response.ok) {
        throw new Error(`Failed to fetch CSV: ${response.statusText}`);
    }
    const csvData = await response.text();

    console.log("Parsing CSV...");
    const records = parse(csvData, {
        columns: true,
        skip_empty_lines: true
    });

    const targetSlugs = new Set<string>();
    for (const row of records) {
        const url = row['URL'] || row['Url'] || row['url'];
        if (url && url.trim()) {
            const trimmedUrl = url.trim();
            try {
                const parsedUrl = new URL(trimmedUrl);
                const pathname = parsedUrl.pathname.replace(/\/$/, '');
                if (pathname) {
                    const slug = pathname.split('/').pop();
                    if (slug) {
                        targetSlugs.add(slug);
                    }
                }
            } catch (e) {
                // Invalid URL, skip
            }
        }
    }

    console.log(`Found ${targetSlugs.size} unique Permapeople URLs in the spreadsheet.`);

    console.log(`Loading ${PLANTS_JSON_PATH}...`);
    const plantsDataRaw = await fs.readFile(PLANTS_JSON_PATH, 'utf-8');
    const plants = JSON.parse(plantsDataRaw);

    const plantMap = new Map<string, any>();
    for (const p of plants) {
        if (p.slug) {
            plantMap.set(p.slug, p);
        }
    }
    console.log(`Loaded ${plantMap.size} plants from ${PLANTS_JSON_PATH}.`);

    try {
        await fs.mkdir(DATA_DIR, { recursive: true });
    } catch (e) {
        // Ignore if exists
    }

    let count = 0;
    for (const slug of targetSlugs) {
        if (plantMap.has(slug)) {
            const plantData = plantMap.get(slug);
            
            // Sort keys in data
            if (plantData.data) {
                const sortedData: any = {};
                Object.keys(plantData.data).sort().forEach(key => {
                    sortedData[key] = plantData.data[key];
                });
                plantData.data = sortedData;
            }

            const outPath = path.join(DATA_DIR, `${slug}.json`);
            await fs.writeFile(outPath, JSON.stringify(plantData, null, 2), 'utf-8');
            count++;
        } else {
            console.log(`Warning: Slug '${slug}' not found in ${PLANTS_JSON_PATH}.`);
        }
    }

    console.log(`Successfully generated ${count} JSON files in the '${DATA_DIR}' directory.`);
}

main().catch(console.error);

File: build_data.py
import csv
import json
import os
import urllib.request
from urllib.parse import urlparse

# Google Sheets CSV Export URL
SHEET_ID = "1fnay67_hc7hU1BRAhbcfnQW5TQGReXGWpn0vIa_dkMY"
CSV_URL = f"https://docs.google.com/spreadsheets/d/{SHEET_ID}/export?format=csv"

# Paths
PLANTS_JSON_PATH = "data/plants.json"
DATA_DIR = "data"

def main():
    print(f"Downloading CSV from {CSV_URL}...")
    req = urllib.request.Request(CSV_URL)
    with urllib.request.urlopen(req) as response:
        csv_data = response.read().decode('utf-8')

    print("Parsing CSV...")
    reader = csv.DictReader(csv_data.splitlines())

    # Extract unique slugs from CSV
    target_slugs = set()
    for row in reader:
        # Search for any column that contains the URL. In main.js it checks 'URL', 'Url', 'url'
        url = row.get('URL') or row.get('Url') or row.get('url')
        if url and url.strip():
            url = url.strip()
            # Extract the last path element as the slug
            parsed = urlparse(url)
            path = parsed.path.rstrip('/')
            if path:
                slug = path.split('/')[-1]
                if slug:
                    target_slugs.add(slug)

    print(f"Found {len(target_slugs)} unique Permapeople URLs in the spreadsheet.")

    print(f"Loading {PLANTS_JSON_PATH}...")
    with open(PLANTS_JSON_PATH, 'r', encoding='utf-8') as f:
        plants = json.load(f)

    # Create a mapping of slug to plant data
    plant_map = {p['slug']: p for p in plants if 'slug' in p}
    print(f"Loaded {len(plant_map)} plants from {PLANTS_JSON_PATH}.")

    # Create output directory
    if not os.path.exists(DATA_DIR):
        os.makedirs(DATA_DIR)

    # Write out JSON files for the targeted slugs
    count = 0
    for slug in target_slugs:
        if slug in plant_map:
            plant_data = plant_map[slug]
            plant_data["data"] = {k: v for k, v in sorted(plant_data["data"].items(), key=lambda x: x[0])}
            out_path = os.path.join(DATA_DIR, f"{slug}.json")
            with open(out_path, 'w', encoding='utf-8') as f:
                json.dump(plant_data, f, indent=2)
            count += 1
        else:
            print(f"Warning: Slug '{slug}' not found in {PLANTS_JSON_PATH}.")

    print(f"Successfully generated {count} JSON files in the '{DATA_DIR}' directory.")

if __name__ == "__main__":
    main()
